File: Python/sortings/k-th_largest_element.py
# Функция основана на алгоритме быстрой сортировки
# Суть задачи следующая: необходимо определить к-ый по величине элемент массива
# Можно, например, отсортировать массив за О(NlogN) и получить к-ый элемент
# А можно за О(N) следующим образом:

def kth_largest_element(arr: list[int], k: int) -> int:
    if len(arr) < 2:
        return arr[0]
    elif len(arr) == 2:
        return arr[k] if arr[0] < arr[1] else arr[::-1][k]
    low = arr[0]
    high = arr[-1]
    medium = arr[len(arr) // 2]
    median = low + high + medium - min(low, high, medium) - max(low, high, medium)
    more = []
    less = []
    equals = []
    for element in arr:
        if element < median:
            less += [element]
        elif element == median:
            equals += [median]
        else:
            more += [element]
    if k < len(less):
        return kth_largest_element(less, k)
    elif k < len(less) + len(equals):
        return equals[0]
    else:
        return kth_largest_element(more, k - len(less) - len(equals))


print(kth_largest_element(list(map(int, input().split())), int(input())))

File: C++/sortings/k-th_largest_element.cpp
#include <vector>
#include <iostream>

// Функция основана на алгоритме быстрой сортировки
// Суть задачи следующая: необходимо определить к-ый по величине элемент массива
// Можно, например, отсортироваить массив за О(NlogN) и получить к-ый элемент
// А можно за О(N) следующим образом:

int kth_largest_element(std::vector<int> v, int k){
    if (v.size() < 2){
        return v[0];
    }
    else if(v.size() == 2){
        if (v[0] > v[1]) std::swap(v[0], v[1]);
        return v[k];
    }
    else{
        std::vector<int> more;
        std::vector<int> less;
        std::vector<int> equals;
        int sz = v.size();
        int low = v.at(0);
        int high = v.at(sz - 1);
        int medium = v.at(sz / 2);
        int median = low + high + medium - std::min(std::min(low, high), medium) - std::max(std::max(low, high), medium);
        for (int i = 0; i < sz; i++){
            int el = v.at(i);
            if (el < median) less.push_back(el);
            else if (el == median) equals.push_back(el);
            else more.push_back(el);
        }
        if (k < less.size()) return kth_largest_element(less, k);
        else if (k < less.size() + equals.size()) return equals[0];
        else return kth_largest_element(more, k - less.size() - equals.size());
    }
}

int main(){
    int t;
    std::cin >> t;
    std::vector<int> v(t);
    for (int i = 0; i < t; i++){
        std::cin >> v[i];
    }
    int k;
    std::cin >> k;
    printf("%d\n", kth_largest_element(v, k));
    return 0;

}